Updates

aguynamedryan opened the pull request:
Here are a few changes I made to the adapter to get it to work for my use cases

aguynamedryan committed:
Add #invalidate_table_schemas

Sequel's implicit qualify feature caches the tables that it finds in
each database in search_path

However, I have a situation where I'm adding new tables to a database
that I still want to implicitly qualify, so I need to invalidate
the cache whenever I create a new table

aguynamedryan committed:
Fix #to_application_timestamp to handle NULL

This method bombed out if the timestamp value was NULL.

jeremyevans commented:
`to_application_timestamp` shouldn't be called with a nil value, that's a bug in the adapter/driver.  Can you try this patch?

```
--- a/lib/rbhive/t_c_l_i_connection.rb
+++ b/lib/rbhive/t_c_l_i_connection.rb
@@ -311,7 +311,9 @@ module RBHive
           vals = row.colVals
           cols.each do |i, col, conv|
             v = vals[i].get_value.value
-            h[col] = conv ? conv[v] : v
+            h[col] = unless v.nil?
+              conv ? conv[v] : v
+            end
           end
           yield h
         end
```

The invalidate_table_schemas method should be fine.

aguynamedryan committed:
Add a series of (temporary?) hacks in DDL methods

After creating a table in Impala, if I immediately queried that table,
most of the time the query would find the table had 0 rows.  BUT if I
waited a few seconds and then ran the query, I'd get rows back.

By adding SYNC_DDL=true, we attempt to ensure that calls to CREATE are
synced across the cluster before the call to Impala returns.

By add "REFRESH" after CREATE, we attempt to make sure the metadata
surrounding the table is appropriately updated before returning
from Sequel#create_table

aguynamedryan committed:
Whitespace fixes

jeremyevans commented:
Sorry, I never got notified about updates to this, I think GitHub doesn't notify you unless you actually post a comment.

For the DDL changes, were you running into problems with all three adapters (impala, rbhive, and jdbc/hive)?  I've noticed race conditions when using the rbhive adapter, but not when using the impala adapter.  If these issues are specific to rbhive, the should be added to the rbhive adapter, not the shared impala adapter.

For the whitespace "fixes", I'm not pedantic about whitespace in any of my projects.  I don't have a problem merging that commit, but there are no guarantees that white space will not be reintroduced.

Have you had a chance to test the diff in my earlier comment?  I think that's a better fix to the application timestamp on nil issue.

jeremyevans committed:
Fix disconnect detection in impala and rbhive adapters

Tested with manually dropping the TCP connection.

aguynamedryan committed:
Revert 80e52e6 and use Jeremy's fix

jeremyevans commented:
@aguynamedryan This looks pretty good, but I haven't heard from you about testing rbhive and impala in regards to the race conditions fixes in commit c035a3ae4d132127e839ad0a285f40ada9ed943f.  I don't want to merge this as is unless this is a problem with both adapters.

aguynamedryan committed:
Remove some debugging lines

aguynamedryan committed:
Use int and bigint for Impala

aguynamedryan committed:
Kerberize impala-ruby

aguynamedryan committed:
Drop SYNC_DDL/refresh calls from shared adapter

jeremyevans commented on lib/sequel/adapters/shared/impala.rb:
```diff
@@ -333,6 +333,16 @@ def metadata_dataset
+      end
+
+      # Impala doesn't like the word "biginteger"
+      def type_literal_generic_bignum(column)
```
This should be `type_literal_generic_bignum_symbol`, at least when Sequel 4.36.0+ is being used.

jeremyevans commented on lib/impala.rb:
```diff
@@ -16,10 +16,10 @@
 module Impala
   DEFAULT_HOST = 'localhost'
   DEFAULT_PORT = 21000
-  class Error < StandardError; end
```
Is there a reason to remove this exception class and make all subclasses inherit from StandardError?

jeremyevans commented:
This changes look good.  Other than the one schema method change mentioned above, it should be ready for commit.

What are your thoughts on moving this adapter to the outlook_insights organization, so you can update it without my input?  I would of course be available to help out with development if you need me.

aguynamedryan committed:
Enable keepalive on impala-ruby sockets

aguynamedryan committed:
Fix up keepalive so it works with Mac OS X

aguynamedryan committed:
Add support for Database#values

Lots of cargo-culting from Postres/Sqlite shared adapters

jeremyevans commented:
@aguynamedryan  The keepalive changes and Database#values look fine.  Have you had a chance to address my concerns noted above about previous commits?

aguynamedryan committed:
Add #compute_stats

aguynamedryan committed:
Add #runtime_profile

jeremyevans commented:
The compute_stats change looks fine.  For runtime_profile, we should probably add a better way to access the profile information.  It looks like the only way to get it currently is to call fetch_rows_and_profile` with an sql string, which is probably not a good API.  Also, it looks like Database#last_profile is not a safe operation, there is no guarantee the cursor still exists outside of the synchronize block inside of execute.

aguynamedryan committed:
Ensure we've pulled payload length from SASL

Seems like sometimes when the client would read the payload, the
length of the returned string was less than expected.  This would
cause the client to misread the length of the incoming payload.

So, so long as the server hasn't returned the full payload length,
we keep reading.  This seems to have fixed the error, hokey as
it may seem.

aguynamedryan committed:
Add more error messages when there is a failure

aguynamedryan committed:
Merge branch 'master' of github.com:aguynamedryan/sequel-impala

jeremyevans commented:
The payload length stuff looks fine to me.  However, it's a bad idea to leave debugging `puts` in production code.

jeremyevans commented:
Closing as the repository has been transfered
